Line Chart 그래프

lacomaco opened the pull request:
## 관련 이슈
Line Chart 만들기 #137 

## 구현 세부 사항
![ezgif-2-fcca77e5a4af](https://user-images.githubusercontent.com/49854357/100513741-dad0f880-31b2-11eb-9906-83ed4f6ad023.gif)
![hover](https://user-images.githubusercontent.com/49854357/100513744-ddcbe900-31b2-11eb-9c18-bfc38631450f.png)

props로 ITransaction 타입의 값들을 넣어주면 자동으로 그래프를 생성하는 Line Chart를 만들었습니다.

꼭 ITransaction 타입 값들로 넣어야 해요

ITransaction 값들중 가장 큰 값을 기준으로 (만) (억) (천억) 과 같은 단위가 먹여지고 가장 큰 값을 기준으로 점의 Y축이 결정됩니다.

그리고 점 좌표가 아래에서 위로 솟구치는 애니메이션과 라인이 그려지는 애니메이션도 넣었습니다. 

좌표내 점을 호버하면 해당 값이 얼마인지 작은 네모칸으로 보여주는 기능도 추가했습니다.

## 작은 공통 컴포넌트 구현

Color Picker와 연동하는 Preview 컴포넌트를 구현했습니다

![화면 캡처 2020-11-29 234559](https://user-images.githubusercontent.com/49854357/100545031-10eaa700-329d-11eb-956a-77fa8591137f.png)

checkbox와 함께 동작하는 dropdown input 컴포넌트를 구현했습니다.

![dropdown](https://user-images.githubusercontent.com/49854357/100545049-311a6600-329d-11eb-93d7-4cb6f3c00022.png)

관리자가 자신이 소유한 가계부를 삭제할 경우 뜨는 모달창을 구현했습니다.

![관리자](https://user-images.githubusercontent.com/49854357/100545069-514a2500-329d-11eb-8882-cf236436feaa.png)


@boostcamp-2020/accountbook_mentor

lacomaco committed:
test : 더미데이터 추가

- Line Chart에 쓸 더미 데이터를 추가하였습니다.

lacomaco committed:
test : line chart 테스트 코드

- line chart 생성시 데이터를 정재하고 필요한 데이터를 만들어 내는 util
  함수들에 대한 코드를 생성하였습니다.

lacomaco committed:
refactor : inputs에 SelectPaymentMethod 컴포넌트 추가

- 실수로 연달아 MultiInputDropDown을 넣었었네요 변경했습니다

lacomaco committed:
refactor : 파일 이동

- graph 폴더에 흩뿌려져 있던 파이 그래프 관련 컴포넌트들을 pie 폴더로
  이동시켰습니다.

lacomaco committed:
feat : #137 라인그래프 스토리북 추가

- 라인그래프를 스토리북에서 볼 수 있도록 추가하였습니다.

lacomaco committed:
feat : #137 라인 그래프 컴포넌트 구현

- 이 컴포넌트에서는 라인 그래프에 쓸 수와 좌표값을 계산하는 역할을
  수행합니다.

- Y축 값의 단위를 계산하여 X축을 생성합니다.
- 시작일과 끝일을 기준으로 Y축에 일별로 텍스트를 찍어냅니다.
- 각 트랜잭션의 값들을 기준으로 Y 좌표를 계산합니다.
- 트랜잭션의 date 값을 기준으로 X좌표를 계산합니다.
- 계산된 트랜잭션들을 전부 모아 계산하여 path을 생성합니다.

lacomaco committed:
feat : lineChartBackground 컴포넌트 생성

- line chart의 x축 y축 경계면을 그리는 컴포넌트입니다.

lacomaco committed:
feat : #137 라인 점 좌표 컴포넌트 생성

- 라인 좌표내 점 좌표를 그리는 컴포넌트를 생성했습니다.
- 호버시 해당 좌표의 값을 좌표 위에 띄어줍니다.
- 아래에서 위로 올라가며 생성되는 애니메이션을 넣었습니다.

lacomaco committed:
feat : #137 라인 차트 텍스트 컴포넌트 생성

- 라인 차트 내에서 쓰이는 텍스트들을 그리는 컴포넌트입니다.

lacomaco committed:
feat : #137 라인 좌표 유틸함수들 생성

- 라인 좌표를 생성하는데 도움을 주는 유틸 함수들을 모아놓은 파일입니다.
- transactions들을 정렬하거나
- 가장 큰 값을 기준으로 X 좌표 비율을 계산하거나
- 일별 Y 좌표 비율을 계산하는등 라인차트 생성의 전반에 쓰이는
  함수들입니다.

lacomaco committed:
feat : #137 X축 생성하는 컴포넌트

- 라인차트를 5등분 하는 X축을 그리는 컴포넌트입니다.

lacomaco committed:
feat : #137 라인 좌표 props Interface

- 라인 차트에 쓰이는 Transaction의 인터페이스를 사용했습니다.
- props로 ITransaction 타입의 값들을 넣어줘야 라인차트를 생성할 수 있어요

lacomaco committed:
chore : react-router-dom 추가

lacomaco committed:
refactor : 테스트를 위해서 생성한 의미없는 함수 삭제

- createPath '안녕ㅎ' 함수를 삭제하였습니다.

mu1616 commented on client/src/components/graph/line-chart/LineChart.tsx:
```diff
@@ -0,0 +1,120 @@
+    return path.getTotalLength();
+  })();
+
+  return (
+    <svg viewBox={`0 0 ${viewX} ${viewY}`}>
+      <LineChartBackground xWidth={viewX} yWidth={viewY} gap={chartGap} />
+      <LineChartText x={0} y={chartGap / 2} value={unit} />
+      {xAxis.map((x, id) => {
+        return <LineChartText key={x} x={0} y={calculateXAxisYPosition(id)} value={x} />;
+      })}
+      {xAxis.map((x, id) => {
+        return <XAxisDraw key={id} y={calculateXAxisYPosition(id)} gap={chartGap} xEnd={viewX} />;
+      })}
+      {yAxis.map((x, id) => {
+        return <LineChartText key={x.position} x={chartGap + x.position} y={viewY - chartGap / 2} value={x.date} />;
+      })}
+
+      {dotsPosition.map((position, id) => {
+        return (
+          <LineChartDot
+            x={position[0]}
+            y={position[1]}
+            key={sortedByDate[id].date.valueOf()}
+            date={sortedByDate[id].date}
+            value={valueFitUnit(sortedByDate[id].value, unit)}
+            yWidth={viewY - 2 * chartGap}
+            gap={chartGap}
+          />
+        );
+      })}
+      <Path d={pathD} length={pathDLength} />
+    </svg>
```
영근님 제가 이해한게 맞는지 확인해주시면 감사하겠습니다!

90line - LineChartBackground : x축과 y축을 그린다.

91line - LineChartText : y축의 단위를 표시한다.

92line - xAxis.map : y축에 숫자(금액)를 표시한다.

95line - xAxis.map : x축에 평행한 수평선을 그린다.

98line - yAxis.map : x축에 숫자(날짜)를 표시한다.

102line - dotPosition.map : 점을 찍는다.

115line - 그래프 선을 그린다.

lacomaco replied:
네 맞습니다 이 해하신거랑 동일하게 동작해요

mu1616 reviewed (approved):
라인그래프 엄청 어려울것 같았는데 하루만에 해치우시다니..  대단하십니다 👍 

제가 svg에 대한 이해가 많이 부족해서 리뷰에 제가 이해한 내용을 작성해봤는데 확인해주시면 감사하겠습니다.
주말에도 열일 하시느라 수고많으셨습니다!!:smile:

lacomaco committed:
test : 가짜 데이터 수정

- 1일부터 시작핮 ㅣ않는 경우도 보기 위해서 수정하였습니다.

lacomaco committed:
refactor : multiInputHOC 리팩토링

- HOC 함수 안에 있던 state을 hooks로 분리하였습니다.
- CheckBox와 함께 쓰이는 dropdown input의 경우 HOC를 활용해서 만드는것은
  도저히 불가능하다고 판단해서
- 값을 관리하는 state을 재활용 하기 위해 hooks로 분리했습니당

lacomaco committed:
feat : 더 많은 경우의 LineChart 를 추가했습니다.

lacomaco committed:
refactor : 리팩토링 코드 추가

- joinWithComma을 재활용 하기 위해 모듈로 export 하였습니다.
- Select Props를 변경햇습니다.

lacomaco committed:
feat : #99 체크박스를 사용하는 input componenet 생성

- 체크박스와 상태를 공유하는 input dropdown component 구현

lacomaco committed:
feat : #91 [공통 컴포넌트 생성] Preview 컴포넌트

- 가계부 표지를 변경하기 전에 미리 확인할 수 있는 Preview 컴포넌트를
  생성했습니다.

lacomaco committed:
feat : 공통 컴포넌트 , 빨강 , 진한 회색 버튼 생성

- 관리자가 소유한 가계부를 삭제할 경우 나오는 모달의 빨강, 진한 회색
  버튼을 생성했습니다.

lacomaco committed:
color : 진한 회색, 진한 빨강색 추가

- 진한 회색과 진한 빨강색을 color constants에 추가하였습니다.

lacomaco committed:
feat : 기획서 4페이지. 모달 창 생성

- 이슈에는 아직 없는 기획서의 4페이지 모달창을 생성했습니다.

- 우선은 상태값이 없기 때문에 글자와 텍스트만 띄우는 정도로
  생성했습니다.

lacomaco committed:
fix : 공통으로 사용되는 buttonProps 분리

- 공통으로 사용되는 buttonProps를 buttonTypes의 ModalCommonButton으로
  분리시켰습니다.

lacomaco committed:
fix : 파일명 오타 변경

- Acoountbook -> Accountbook으로 변경하였습니다.

mu1616 commented on client/src/hook/use-drop-down/useMultiDropDown.tsx:
```diff
@@ -0,0 +1,41 @@
+    } else {
+      selected.add(value);
+    }
+    setSelected(setImmutable(selected));
```
이 부분은 어떤 역할인가요..??

lacomaco replied:
selectChange 함수는 값을 변경시킬때 호출되는 함수인데요 

35번 라인은 최종적으로 변경된 값을 selected state에 반영하는 역할입니다. 

selected state는 Set 자료구조 이기 때문에 불변성을 유지하기 위해 setImmutable 함수를 통해서 완전히 새로운 Set을 state에 넣는 코드에요

mu1616 replied:
아하 감사합니다!

Mong-Gu reviewed (approved):
대단하십니다 정말... 열심히 보면서 공부하겠습니다 하루아침만에 보고 끝내기엔 저에게 너무 버거운 코드네요 ㅋㅋㅋ ㅜㅜㅜㅜ 그리고 lint가 혹시 적용 안 되시는 걸까요? 다음부터 lint도 적용해주시면 감사하겠습니다!

lacomaco committed:
fix : 의미없는 console.log 삭제

- empty-arrow-function lint를 피하기 위해 추가했던 console.log를
  삭제하고

- 주석을 채워 넣었습니다.

bbbyung2 commented on client/src/components/common/preview/Preview.tsx:
```diff
@@ -0,0 +1,46 @@
+`;
+
+const Preview: React.FC<PreviewProps> = ({ title, description, color, onChange }: PreviewProps) => {
+  console.log(color);
```
앗.. 이것만 삭제해주시면 좋을 것 같습니다!👍🏻

lacomaco replied:
아앗 감사합니다

lacomaco committed:
fix : 의미없는 console.log 삭제

- Preview.tsx 파일에서 console.log를 삭제하였습니다.
